### PR TITLE
Add 'Delete by Name' Functionality and Enhance Test Coverage

### PR DESCRIPTION
**Pull request by wapisai**
This PR introduces the ability to delete items by name, addressing a key user request for enhanced flexibility in the delete command. Additionally, we've expanded the test coverage to ensure the reliability and robustness of this new feature.

## Changes Made:
- Modified `DeleteCommand` and `DeleteCommandParser` to support deletion by name, allowing for more intuitive user interactions.
- Added comprehensive unit tests in DeleteCommandTest to cover new scenarios introduced by the 'delete by name' functionality, ensuring robustness and error handling.

## Testing:
- Implemented new unit tests focusing on edge cases and typical usage scenarios to validate the correctness of the delete operation.
- Conducted manual testing to ensure integration with existing functionalities remains seamless.

## Fixes and Related Issues:
This PR addresses the issue #67 

## Code Quality and Compatibility:
- Ensured compatibility with existing commands and functionalities, verified through automated regression testing.
- Conducted code quality checks to maintain consistency and adherence to project standards.

## Additional Notes:
- Priority for this feature has been set to Medium, reflecting its importance to user experience.
- We encourage further review and testing from the community to refine and validate the new tests and functionalities introduced.

## Please feel free to provide feedback or suggestions for further improvements!

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/DeleteCommandParser.java [hunk 1]
```diff
@@ -18,12 +19,16 @@ public class DeleteCommandParser implements Parser<DeleteCommand> {
...
+                Name name = ParserUtil.parseName(args.replace("n/", "").trim());
+                return new DeleteCommand(name);
+            } else {
+                Index index = ParserUtil.parseIndex(args);
+                return new DeleteCommand(index);
+            }
         } catch (ParseException pe) {
             throw new ParseException(
-                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE), pe);
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE));
```
Comment (appleraincoat) [inline src/main/java/seedu/address/logic/parser/DeleteCommandParser.java:31]:
```suggestion
                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE), pe);
```

Comment (appleraincoat) [reply src/main/java/seedu/address/logic/parser/DeleteCommandParser.java:31]: Perhaps you can consider including the ParseException pe as there might be extra information contained within that needs to output in the exception message.

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 2]
```diff
@@ -41,6 +42,20 @@ public void execute_validIndexUnfilteredList_success() {
...
+
+        String expectedMessage = String.format(DeleteCommand.MESSAGE_DELETE_PERSON_SUCCESS,
+                Messages.format(personToDelete));
+
+        ModelManager expectedModel = new ModelManager(model.getAddressBook(), new UserPrefs());
+        expectedModel.deletePerson(personToDelete);
+
+        assertCommandSuccess(deleteCommand, model, expectedMessage, expectedModel);
+    }
+
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:58]: Test looks good to me!

**Review by appleraincoat — COMMENTED** [review]
Looks good to me, apart from the minor issue with the parse exception :)

File: src/main/java/seedu/address/logic/parser/DeleteCommandParser.java [hunk 3]
```diff
@@ -18,12 +19,16 @@ public class DeleteCommandParser implements Parser<DeleteCommand> {
...
         try {
-            Index index = ParserUtil.parseIndex(args);
-            return new DeleteCommand(index);
+            if (args.contains("n/")) {
+                Name name = ParserUtil.parseName(args.replace("n/", "").trim());
+                return new DeleteCommand(name);
+            } else {
+                Index index = ParserUtil.parseIndex(args);
+                return new DeleteCommand(index);
+            }
```
Comment (UdhayaShan1) [inline src/main/java/seedu/address/logic/parser/DeleteCommandParser.java:28]: Try utilise multiMap to detect n/ instead, remove the try catch no need to catch the parseexception using this too, this also shows exact message of empty name is user just types `delete n/`
```suggestion
ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME);
if (arePrefixesPresent(PREFIX_NAME)) {
Name name = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get());
return new DeleteCommand(name);
} else {
//....
}

private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
    return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
}
                
```

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 4]
```diff
@@ -19,35 +21,68 @@ public class DeleteCommand extends Command {
     public static final String COMMAND_WORD = "delete";
 
     public static final String MESSAGE_USAGE = COMMAND_WORD
-            + ": Deletes the client identified by the index number used in the displayed client list.\n"
-            + "Parameters: INDEX (must be a positive integer)\n"
-            + "Example: " + COMMAND_WORD + " 1";
+            + ": Deletes the client identified by the index number or name used in the displayed client list.\n"
```
Comment (UdhayaShan1) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:24]:
```suggestion
            + ": Deletes the client identified by the index number or full name used in the displayed client list.\n"
```

Comment (UdhayaShan1) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:24]: Important to emphasise full name as it may be passed off as a bug

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 5]
```diff
@@ -19,35 +21,68 @@ public class DeleteCommand extends Command {
...
+
+    private Person deleteByName(List<Person> lastShownList) throws CommandException {
+        for (Person person : lastShownList) {
+            if (person.getName().equals(targetName)) {
+                return person;
+            }
+        }
+        throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_NAME);
+    }
+
```
Comment (UdhayaShan1) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:85]: Can consider doing this in another class like `DeleteCommandExecutor` meant just for deletion logic

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
LGTM mostly, just some SLAP issues

## Pls dont do on master 😿

**Comment by wapisai** [comment]
> LGTM mostly, just some SLAP issues
> 
> ## Please dont do on master 😿

**Comment by wapisai** [comment]
> LGTM mostly, just some SLAP issues
> 
> ## Please dont do on master 😿

I won't after this pr 😭

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 6]
```diff
@@ -19,35 +21,68 @@ public class DeleteCommand extends Command {
     public static final String COMMAND_WORD = "delete";
 
     public static final String MESSAGE_USAGE = COMMAND_WORD
-            + ": Deletes the client identified by the index number used in the displayed client list.\n"
-            + "Parameters: INDEX (must be a positive integer)\n"
-            + "Example: " + COMMAND_WORD + " 1";
+            + ": Deletes the client identified by the index number or full name used in the displayed client list.\n"
```
Comment (jyztintan) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:24]: good and detailed addition 👍

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 5 above)
Comment (jyztintan) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:85]: agree, SLAP

File: src/main/java/seedu/address/logic/parser/DeleteCommandParser.java
Comment (jyztintan) [inline src/main/java/seedu/address/logic/parser/DeleteCommandParser.java:1]: Tests looks good 👍

**Review by jyztintan — APPROVED** [review]
Thanks for adding this new delete by full name feature!

Think in future can consider some test cases just to showcase its behaviour:
Only delete by exact full name, and not by:
- Matching other fields (`delete Sydney` should not delete someone with _Address_ `Sydney`
- Substring (`delete Udhaya` should not delete someone with name `Udhaya Shanmugam`

**Review by UdhayaShan1 — APPROVED** [review]

**Comment by wapisai** [comment]
> Thanks for adding this new delete by full name feature!
> 
> Think in future can consider some test cases just to showcase its behaviour: Only delete by exact full name, and not by:
> 
> * Matching other fields (`delete Sydney` should not delete someone with _Address_ `Sydney`
> * Substring (`delete Udhaya` should not delete someone with name `Udhaya Shanmugam`

Thanks! Added

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 5 above)
Comment (wapisai) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java]: Done!

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 7]
```diff
@@ -124,6 +124,41 @@ public void equals() {
...
     }
 
+    @Test
+    public void equalsName() {
+        DeleteCommand deleteUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya").build().getName());
+        DeleteCommand deleteNotUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("not Udhaya").build().getName());
+        DeleteCommand deleteUdhayaShanugamCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya Shanugam").build().getName());
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:134]: image

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 8]
```diff
@@ -49,6 +64,14 @@ public void execute_invalidIndexUnfilteredList_throwsCommandException() {
         assertCommandFailure(deleteCommand, model, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
     }
 
+    @Test
+    public void execute_invalidNameUnfilteredList_throwsCommandException() {
+        Person personToDelete = new PersonBuilder().withName("b").build();
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:69]: Good!

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 9]
```diff
@@ -101,6 +124,41 @@ public void equals() {
...
+        assertFalse(deleteUdhayaCommand.equals(null));
+
+        // different person -> returns false
+        assertFalse(deleteUdhayaCommand.equals(deleteNotUdhayaCommand));
+
+        // delete udhaya should not delete someone with address udhaya
+        assertFalse(deleteUdhayaCommand.equals(deleteAddressCommand));
+
+        //delete Udhaya should not delete someone with name Udhaya Shanugam
+        assertFalse(deleteUdhayaCommand.equals(deleteUdhayaShanugamCommand));
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:159]: `equalsName()` LGTM

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 10]
```diff
@@ -101,6 +124,41 @@ public void equals() {
...
     }
 
+    @Test
+    public void equalsName() {
+        DeleteCommand deleteUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya").build().getName());
+        DeleteCommand deleteNotUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("not Udhaya").build().getName());
+        DeleteCommand deleteUdhayaShanugamCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya Shanugam").build().getName());
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:134]:
```suggestion
                new PersonBuilder().withName("Udhaya Shanmugam").build().getName());
```
😭

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 11]
```diff
@@ -124,6 +124,41 @@ public void equals() {
...
+        assertFalse(deleteUdhayaCommand.equals(null));
+
+        // different person -> returns false
+        assertFalse(deleteUdhayaCommand.equals(deleteNotUdhayaCommand));
+
+        // delete udhaya should not delete someone with address udhaya
+        assertFalse(deleteUdhayaCommand.equals(deleteAddressCommand));
+
+        //delete Udhaya should not delete someone with name Udhaya Shanugam
+        assertFalse(deleteUdhayaCommand.equals(deleteUdhayaShanugamCommand));
```
Comment (jyztintan) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:159]: image

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java
Comment (jyztintan) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:1]: Thanks for adding the new test cases! They look good!

**Review by jyztintan — APPROVED** [review]
LGTM!

**Review by UdhayaShan1 — APPROVED** [review]
LGTM!!!!

**Review by appleraincoat — APPROVED** [review]
Issues have already been brought up by others
LGTM!
